services/src/main/java/org/keycloak/protocol/oid4vc/LocatableProvider.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright 2024 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.protocol.oid4vc;
+
+import org.keycloak.provider.Provider;
+
+/**
+ * @author <a href="mailto:Ingrid.Kamga@adorsys.com">Ingrid Kamga</a>
+ */
+public interface LocatableProvider extends Provider {
```

Comment:
Can you elaborate a bit why this abstraction is needed?

Reply:
Thank you for raising this concern. This abstraction was mainly introduced to reuse the pre-existing logic for loading signing services. Using their `locator` methods, these components compute their identifiers in a way that cannot be achieved by their factories. For example, a signing service is searchable by a combination of a credential format, type, and config ID.

https://github.com/keycloak/keycloak/blob/faf36dd2f061fe0c0392b5b731f3fdaf7c33c34b/services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/VerifiableCredentialsSigningService.java#L50-L63

Although this logic is not strictly required for credential builders (subsequent to the iterated changes in their development), it appears to be a valid workaround for loading ComponentModel-configurable providers.

Reply:
Thanks for clarifiying that. So in essence you are looking for a way to narrow down a concrete provider among multiple provider implementations of the same type that provide some specialization.

Can't you leverage the existing lookup by "providertype and providerId" functionality? You could have multiple provider implementations, all with a different providerId. The providerId could be derived from some parameters, in your example: "format::type::configId", or `providerId.variant1` as the type is already determined by the provider interface.

In your code you would then just compute the appropriate "providerId" dynamically:

```
var concreteProviderId = "providerId." + "variant1";
var provider = session.getProvider(MyProvider.class, concreteProviderId);
```

This works if you register all providers during build / startup. Would that work for you? With this we wouldn't need to introduce a new "locator" mechanism to the keycloak core.

Reply:
Thank you for providing further guidance.

> The providerId could be derived from some parameters.

The precise problem is that if these parameters originate from `ComponentModel` configurations, then we likely can't channel them to the `getId` redefinition in the factory.

Second, it appears that `session.getProvider` does not work if you have `ComponentFactory`'s instead of plain `ProviderFactory`'s.

Regardless, I could rewrite the code without introducing the problematic abstraction. Note that I also reverted the way signing services were previously loaded. Please could you have a look?